### DISCUSSION TITLE
Implicit transaction and rollback errors

### DISCUSSION THREAD

**Discussion by Arcus16** [mikro-orm/mikro-orm#3398] [Q&A] [answered]
Hi,
I have end point when I want to create serval instances of enitites.  I have this code 
```javascript
        try {
            await this.entityManager.begin();
            await this.resourceService.create(input, user);
            await this.entityManager.commit();
        } catch (e: unknown) {
            await this.entityManager.rollback();
            throw e;
        }
```
And it works (I will get an error if one occures).
I am trying to use implict transaction like this:
```javascript
        try {
            await this.resourceService.create(input, user);
            await this.entityManager.flush();
        } catch (e: unknown) {
            console.error(e);
            throw e;
        }
```
and everything is working when there is no error, but when error occurs transaction is rollbacked but no error will be throwed. I want to ask if this is intended behavior?

**Comment by B4nan** [comment]
Share more details, this is too generic question.

**Comment by Arcus16** [comment, marked as answer]
Hi,
I debugged my code and I find out the cause. In function `create` I create meny entities by calling other services/fucntions and one function calls `persistAndFlush`. So what is happening is this:

1. Create entity X
2. Create entity Y (X and Y are related,  X has many Y (1:M))
3. persistAndFlush Y 
4. persist entity X
5. Flush evrything (flush X)

I was able to fix my code when I changed the `persistAndFlush` to `persist`.

**Reply by B4nan** [reply]
Just note that if `Y` is related to `X` (it is reachable from it), it will be flushed together with the first `persistAndFlush` in point 3. Sounds like you are doing some parallel work in there, otherwise this shouldnt really matter - still, you should aim to flush only once, so your fix is correct approach.
